plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java

```diff
@@ -339,6 +346,87 @@ public void testOptimizingPartitionsOfV2TableWithGlobalEqualityDeleteFile()
     }
 
+    @Test(timeOut = 90_000)
+    public void testOptimizeDuringWriteOperations()
+            throws Exception
+    {
+        int invocationCount = 3;
+        boolean optimizeSucceeded = false;
+        for (int attempt = 0; attempt < invocationCount; attempt++) {
+            optimizeSucceeded = runOptimizeDuringWriteOperations();
+            if (optimizeSucceeded) {
+                break;
+            }
+        }
+        assertTrue(optimizeSucceeded, "Expected optimize to succeed in at least one test attempt");
+    }
+
+    private boolean runOptimizeDuringWriteOperations()
+            throws Exception
+    {
+        int threads = 5;
+        int deletionThreads = threads - 1;
+        int rows = 100;
+        CyclicBarrier barrier = new CyclicBarrier(threads);
+        ExecutorService executor = newFixedThreadPool(threads);
+        try (TestTable table = new TestTable(
+                getQueryRunner()::execute,
+                "test_optimize_during_write_operations",
+                "(int_col INT)")) {
+            String tableName = table.getName();
+            for (int i = 0; i < rows; i++) {
+                assertUpdate(format("INSERT INTO %s VALUES %s", tableName, i), 1);
+            }
+
+            int rowsPerThread = rows / deletionThreads;
+            List<Future<List<Boolean>>> deletionFutures = IntStream.range(0, deletionThreads)
+                    .mapToObj(threadNumber -> executor.submit(() -> {
+                        barrier.await(10, SECONDS);
+                        List<Boolean> successfulDeletes = new ArrayList<>();
+                        for (int i = 0; i < rowsPerThread; i++) {
+                            try {
+                                int rowNumber = threadNumber * rowsPerThread + i;
+                                getQueryRunner().execute(format("DELETE FROM %s WHERE int_col = %s", tableName, rowNumber));
+                                successfulDeletes.add(true);
+                            }
+                            catch (RuntimeException e) {
+                                successfulDeletes.add(false);
+                            }
+                        }
+                        return successfulDeletes;
+                    }))
+                    .collect(toImmutableList());
+
+            Future<Boolean> optimizeFuture = executor.submit(() -> {
+                try {
+                    barrier.await(10, SECONDS);
+                    assertUpdate("ALTER TABLE %s EXECUTE optimize".formatted(tableName));
+                    return true;
+                }
+                catch (Exception e) {
+                    return false;
+                }
+            });
+
+            int expectedRows = 0;
+            for (Future<List<Boolean>> future : deletionFutures) {
+                for (Boolean successfulDelete : future.get()) {
+                    if (!successfulDelete) {
+                        expectedRows++;
```

Comment:
it may be that `rowsPerThread * deletionThreads != rows` (if i change threads not to be a divisor of `rows` + 1)

also,  a delete thread

so
```
int expectedRows = rows;
... 
    if (successfulDelete) {
       expectedRows --;
    }
```